Allow threadsafe access to buffer of type inference profiling trees

vilterp opened the pull request:
This PR allows accessing the results of type inference profiling (via `@snoopi_deep`) from a separate background thread.

Note that type inference on _multiple threads_ while snooping is enabled is **still not threadsafe** as of this PR, since starting in 1.9 type inference can happen in parallel on multiple threads. We will fix this in a follow-up PR.

In the meantime, this is a step in the right direction, and fixes thread safety for the specific case of one profiling thread and one inference thread, and this PR could be backported to 1.8.*.

TODO:
- ~print deprecation warning (`println` and `@warn` are undefined in the compiler)~
    - decided not to do this; it's infeasible and it's a compiler internal.

vilterp committed:
Allow threadsafe access to buffer ot type inference profiling trees

Co-authored-by: Nathan Daly <NHDaly@gmail.com>

pchintalapudi commented on base/compiler/typeinfer.jl:
```diff
@@ -66,6 +66,36 @@ Timing(mi_info, start_time) = Timing(mi_info, start_time, start_time, UInt64(0),
+
+Empty out the previously recorded type inference timings (`Core.Compiler._timings`).
+"""
+function clear_and_fetch_timings()
+    ccall(:jl_typeinf_profiling_lock_begin, Cvoid, ())
+    try
+        timings_copy = copy(_finished_timings)
+        empty!(_finished_timings)
+        return timings_copy
+    finally
+        ccall(:jl_typeinf_profiling_lock_end, Cvoid, ())
+    end
+    return nothing
+end
+
+function finish_timing_profile(timing::Timing)
+    ccall(:jl_typeinf_profiling_lock_begin, Cvoid, ())
+    try
+        push!(_finished_timings, timing)
+    finally
+        ccall(:jl_typeinf_profiling_lock_end, Cvoid, ())
+    end
+end
```
These functions should be implemented in C, and in general we should not take C locks from Julia-level type inference.

NHDaly replied:
@pchintalapudi: See this discussion on slack where @vtjnash suggested we use a new C lock for this:
https://julialang.slack.com/archives/C688QKS7Q/p1668711395201969?thread_ts=1668710990.816479&cid=C688QKS7Q

Why would it be preferable to implement in C? I don't see any benefit to that.

pchintalapudi replied:
The reasoning is actually that there's a deadlock where thread A holds `typeinf_profiling_lock` and thread B holds the codegen lock, but thread A decides it needs to codegen `push!`, which requires holding the codegen lock, and thread B decides it needs to type infer something else, calling typeinf, and then blocks waiting for `typeinf_profiling_lock`.

Alternatively, we can look at it from the locking documentation perspective: https://github.com/JuliaLang/julia/blob/master/doc/src/devdocs/locks.md. `typeinf_profililng_lock` should be a level 2 lock since it guards a push that may GC, but julia code may not be called while holding anything under a level 6 lock. The only way to satisfy this criterion is to implement the method in C, which also avoids the deadlock since it's guaranteed that C will not trigger type inference or codegen.

pchintalapudi replied:
Yeah I wouldn't recommend trying that mostly because this is type inference and not user code, so the deadlock could happen anytime. If the lock was outside type inference, then it wouldn't matter and it could be written in Julia.

NHDaly replied:
yep, makes sense. Thanks a bunch! I've moved it to C (C++) in the latest commit. I'll also add this lock to that doc. 👍

NHDaly replied:
👍 k, added to the list as well. Thank you for the quick and careful review, @pchintalapudi! 👍 ❤️

NHDaly commented on base/compiler/typeinfer.jl:
```diff
@@ -74,59 +104,46 @@ _time_ns() = ccall(:jl_hrtime, UInt64, ())  # Re-implemented here because Base n
-    end
-    return nothing
-end
+# """
+#     Core.Compiler.reset_timings()
+# 
+# Empty out the previously recorded type inference timings (`Core.Compiler._timings`), and
+# start the ROOT() timer again. `ROOT()` measures all time spent _outside_ inference.
+# """
+# function reset_timings()
+#     empty!(_timings)
+#     push!(_timings, Timing(
+#         # The MethodInstance for ROOT(), and default empty values for other fields.
+#         InferenceFrameInfo(ROOTmi, 0x0, Any[], Any[Core.Const(ROOT)], 1),
+#         _time_ns()))
+#     return nothing
+# end
+# reset_timings()
```
We can delete this now.
```suggestion
```

vilterp commented:
Patch on 1.8: https://github.com/JuliaLang/julia/compare/release-1.8...vilterp:julia:pv-inference-timings-v1.8.2?expand=1

vilterp commented:
Mysterious error in the tests:
```
Package EnglishText errored during testing
  | Stacktrace:
  | [1] pkgerror(msg::String)
  | @ Pkg.Types /cache/build/default-amdci5-0/julialang/julia-master/julia-959ef6c00a/share/julia/stdlib/v1.10/Pkg/src/Types.jl:68
  | [2] test(ctx::Pkg.Types.Context, pkgs::Vector{Pkg.Types.PackageSpec}; coverage::Bool, julia_args::Cmd, test_args::Cmd, test_fn::Nothing, force_latest_compatible_version::Bool, allow_earlier_backwards_compatible_versions::Bool, allow_reresolve::Bool)
  | @ Pkg.Operations /cache/build/default-amdci5-0/julialang/julia-master/julia-959ef6c00a/share/julia/stdlib/v1.10/Pkg/src/Operations.jl:1878
  | [3] test
```

Can't find the actual error message. Anyone know what this is?

NHDaly commented:
Looks like that failing test is coming from here:
https://github.com/JuliaLang/Pkg.jl/blob/d144b12935c56c0a0fc2f730e2960ae02096f131/test/new.jl#L301-L306

I guess the julia tests must pull in the Pkg.jl tests. And _those_ tests are cloning and testing the [EnglishText.jl package](https://github.com/TotalVerb/EnglishText.jl) 😅 

It seems _probably_ unrelated, although it's a little weird that those same tests keep failing on multiple commits.. 🤔

NHDaly commented:
I tried pulling down `EnglishText` locally to see why it's failing, and it turns out that it's because EnglishText transitively depends on `SnoopCompile`, using `@snoopi_deep` through `SnoopPrecompile` to precompile a package!
So we need to merge our fixes to SnoopCompile to update it for this PR _first_, before these tests will pass, I think!

Pretty wild that SnoopCompile is part of julia build tests like this. But also very cool!

vchuravy commented on base/compiler/typeinfer.jl:
```diff
@@ -97,16 +100,49 @@ const _finished_timings = Timing[]
+# *not* in type inference during a given recording trace. It is used as a "dummy" node.
+function ROOT() end
+const ROOTmi = Core.Compiler.specialize_method(
+    first(Core.Compiler.methods(ROOT)), Tuple{typeof(ROOT)}, Core.svec())
```
```suggestion
    first(Core.Compiler.methods(ROOT)), Tuple{typeof(ROOT)}, Core.svec())

```

vchuravy commented on base/compiler/typeinfer.jl:
```diff
@@ -97,16 +100,49 @@ const _finished_timings = Timing[]
+const ROOTmi = Core.Compiler.specialize_method(
+    first(Core.Compiler.methods(ROOT)), Tuple{typeof(ROOT)}, Core.svec())
+"""
+    Core.Compiler.reset_timings()
```
```suggestion
    Core.Compiler.reset_timings()

```

NHDaly commented:
>  print deprecation warning (`println` and `@warn` are undefined in the compiler)

@vchuravy or @pchintalapudi: do you have any advice on how best to mark the old functions as deprecated? I was thinking about possibly moving the implementations to `base/deprecated.jl`, and `@eval`ing them back into `Core.Compiler.Timings`. This way, we can interpolate in the deprecation utilities.

Does that seem like the right approach to you?

NHDaly commented:
K, thanks, that's helpful. 👍

NHDaly commented:
Maybe we can still log a deprecation warning via `Core.println()`, just for good measure. we have a PR out to update SnoopCompile to the new API as well.

NHDaly commented on src/inference-profiling.cpp:
```diff
@@ -0,0 +1,48 @@
+// Implementation for type inference profiling
```
it started out using a C++ vector, but yeah agreed, could just be a C file now. 👍

vilterp replied:
Done, thx @NHDaly 🙏

vilterp replied:
this appears to be `.cpp` again… Did something happen to your commit @NHDaly? 🤔

NHDaly replied:
thx

NHDaly commented:
👍 let us know when this is good for final review

NHDaly commented on base/compiler/typeinfer.jl:
```diff
@@ -66,6 +69,28 @@ Timing(mi_info, start_time) = Timing(mi_info, start_time, start_time, UInt64(0),
+    ccall(:jl_typeinf_profiling_push_timing, Cvoid, (Any, Any,), _finished_timings, timing)
+end
+
+# TODO(PR): Comment this vector, explain that it should only be accessed thru C.
```
```suggestion
# DO NOT ACCESS DIRECTLY. This vector should only be accessed through the 
# functions above. It is a buffer that lives in the Julia module only to be *rooted*
# for GC, but all accesses to the vector must go through C code, in order to be
# thread safe.
```

vchuravy replied:
Bump?

NHDaly committed:
Make inference profiling thread safe for julia 1.9+

Starting in 1.9, julia does not hold the codegen lock for all of
typeinference! This means that we cannot use a single shared
scratch-space for profiling type inference.

This PR changes the scratch space to use Task Local Storage, so that we
build a separate inference profile tree per Task doing inference, and
then report them to the global results buffer at the end.

vilterp commented:
Yep, this is ready for review, except that I'm seeing a lot of errors like

```
Expression: resp isa Response
  | Evaluated: Downloads.RequestError("https://untrusted-root.badssl.julialang.org", 28, "Failed to connect to untrusted-root.badssl.julialang.org port 443 after 15016 ms: Operation timed out", Downloads.Response(nothing, "https://untrusted-root.badssl.julialang.org/", 0, "", Pair{String, String}[])) isa Downloads.Response
  | Error in testset Downloads:
  | Error During Test at /cache/build/default-amdci5-3/julialang/julia-master/julia-b34a42122c/share/julia/stdlib/v1.10/Downloads/test/runtests.jl:515
  | Test threw exception
  | Expression: resp.status == 200
  | type RequestError has no field status
```

Rebasing on master just in case that fixes it

NHDaly commented:
Thanks! Will do a final review pass

NHDaly commented on base/compiler/typeinfer.jl:
```diff
@@ -66,6 +69,28 @@ Timing(mi_info, start_time) = Timing(mi_info, start_time, start_time, UInt64(0),
+"""
+    Core.Compiler.Timings.clear_and_fetch_timings()
+
+Return, then clear, the previously recorded type inference timings (`Core.Compiler._timings`).
```
```suggestion
Return, then clear, the previously recorded type inference timings.
```
This was not correct anymore, and also we don't want people to be accessing the correct vector (`_finished_timings`) by name anyway.

vchuravy commented on base/compiler/typeinfer.jl:
```diff
@@ -74,16 +99,20 @@ _time_ns() = ccall(:jl_hrtime, UInt64, ())  # Re-implemented here because Base n
 Empty out the previously recorded type inference timings (`Core.Compiler._timings`), and
 start the ROOT() timer again. `ROOT()` measures all time spent _outside_ inference.
+
+DEPRECATED: this will be removed; use `clear_and_fetch_timings` instead.
```
```suggestion
!!! info
    This functions is deprecated as of Julia 1.9; use `clear_and_fetch_timings` instead.
```

NHDaly replied:
thx

vchuravy commented on base/compiler/typeinfer.jl:
```diff
@@ -105,44 +133,31 @@ reset_timings()
-            parent_timer.children,
-            parent_timer.bt,
-        )
+        _timings[end].time += accum_time
```
Should this be an atomic increment?

NHDaly replied:
it doesn't need to be, since it's essentially a thread-local object since (as of julia 1.8) this is guarded by the codgen lock.

Indeed it's not thread safe starting in 1.9, that's what https://github.com/JuliaLang/julia/pull/48051 is for.
We were going to merge these separately so that we can backport this to 1.8 in our own build.

vchuravy replied:
The right strategy then is to have them be separate commits in the same PR.

vilterp replied:
Hm, so you think we should move that PR's changes into this one before merging this one? 🤔 

I just squashed this one down to one commit to make that easier; can also make this into an atomic

vilterp replied:
Hm, I tried this just now but I'm not allowed to `using Base.Threads: Atomic, atomic_add!` in the compiler, since it's not defined yet 🤔

NHDaly replied:
i'm fine with doing it in one PR. But note that this is already broken without this PR, and so doing it in two incremental steps didn't seem worse than doing it in one step, and that way we could merge this one which is already pretty much done...

But we can merge them both into here if that's easier, yeah.

NHDaly replied:
Okay, just reviewed the commit in the thread safety PR again (https://github.com/JuliaLang/julia/pull/48051/commits/18f30c8b6da7aff311e5a5dfc17799cc6c67bde8), and I don't think that it's finished. I don't think we tested it at all, and we certainly haven't tested it in our codebase yet.

I would not feel comfortable ramming that in yet, BUT i would really like us to merge this PR to close this loop and prevent it from bitrotting.

Since this doesn't make the thread safety any _worse_, let's merge this PR now, and then we can come back to the thread safety discussion in https://github.com/JuliaLang/julia/pull/48051.

vchuravy replied:
> Maybe we can't use atomics and need to rely on a lock?

`@atomic` not `Base.Threads.Atomic` to mark a field as atomic.

NHDaly replied:
Okay, i merged in https://github.com/JuliaLang/julia/pull/48051. We're testing this now in our 1.9 branch at RAI, and assuming it looks okay then i'll feel good about moving forward with this.

I guess it's also got its own tests in this PR, so it's probably fine too.

vilterp replied:
Thanks 🙏

NHDaly reviewed (approved):
Thanks @vilterp, this LGTM!

Can you address the last open comments? Then this looks good to merge

NHDaly commented on base/compiler/typeinfer.jl:
```diff
@@ -66,24 +69,70 @@ Timing(mi_info, start_time) = Timing(mi_info, start_time, start_time, UInt64(0),
+    if t.storage === nothing
+        t.storage = IdDict()
+    end
+    return (t.storage)::IdDict{Any,Any}
```
@vchuravy / @pchintalapudi / @kpamnany: I finally got around to testing this, and indeed it doesn't work. :(

It turns out that `Base.IdDict` and `Core.Compiler.IdDict` are not the same thing. 😢 

Does anyone know why that is? Why doesn't Base just reexport the IdDict from Core.Compiler?
I.e. here: https://github.com/JuliaLang/julia/blob/80aeebe032a89455482134e8a9f1d46b1f5f2b40/base/Base.jl#L165
Why doesn't it just reexport the IdDict from Core.Compiler?

NHDaly replied:
BUT anyway, the error we encounter is that we're not allowed to set the Task's `.storage` field to a Core.Compiler.IdDict; it's expected to be a Base.IdDict.

So is there just no way to use task-local storage from Core.Compiler?

NHDaly replied:
Possibly a more robust solution here, which might also last longer if we parallelize inference, would be to store this on the `AbstractInterpreter` or `InferenceState` objects? Does anyone know if that's plausible? We essentially want some object that lives for the lifetime of the inference invocation, and is local to this invocation, where we can store the stack of timers.

NHDaly replied:
Or we'd have to thread it through every function call, but that seems like possibly a nonstarter.

vchuravy replied:
> So is there just no way to use task-local storage from Core.Compiler?

Nobody needed this before it seems :) 

> Does anyone know why that is? Why doesn't Base just reexport the IdDict from Core.Compiler?

Likely to isolate the compiler code from Base.

vchuravy replied:
You can do:

```
if isdefined(Core, :Main)
  Core.Main.Base.get_task_tls(current_task())
end
```

to get around the bootstrapping issue.

vtjnash replied:
It is because the compiler is forbidden from accessing TLS. Since we hijack the current task to run codegen, it could corrupt the running task if it does anything to interact with it.

vtjnash replied:
(instead, you could perhaps replace the TLS on entry with one from Core.Compiler, and restore it on return)

NHDaly commented:
This still has `backport-1.9`, but we never managed to finish it up. :(

I think we shouldn't hold up 1.9 on this. Should I remove the backport label?  CC: @vchuravy

It means that users will need to be careful not to run `@snoopi_deep` on a process with multiple threads.. :/ Maybe we should at least backport a warning to the docs?

DilumAluthge commented:
Okay, so this PR was opened in November 2022, and it was approved by @jpsamaroo in January 2023. However, since that approval, I see some comments from Julian suggesting that the PR doesn't actually work.

Additionally, there has been no activity on this PR for two years+.

@jpsamaroo @vilterp What is the status of this PR? Are y'all still working on it?
